src/engine.rs

```diff
@@ -426,6 +462,151 @@ impl Engine {
             .get_raw(account_id)
             .map(|result| Address(result.as_slice().try_into().unwrap()))
     }
+
+    pub fn register_token(&mut self, erc20_token: &[u8], nep141_token: &[u8]) {
+        // Check that this nep141 token was not registered before, they can only be registered once.
+        let map = Self::nep141_erc20_map();
+        assert!(map.lookup_left(nep141_token).is_none());
+        map.insert(nep141_token, erc20_token);
+    }
+
+    pub fn get_erc20_from_nep141(&self, nep141_token: &[u8]) -> Option<Vec<u8>> {
+        Self::nep141_erc20_map().lookup_left(nep141_token)
+    }
+
+    /// Transfers an amount from a given sender to a receiver, provided that
+    /// the have enough in their balance.
+    ///
+    /// If the sender can send, and the receiver can receive, then the transfer
+    /// will execute successfully.
+    pub fn transfer(
+        &mut self,
+        sender: Address,
+        receiver: Address,
+        value: Wei,
+        gas_limit: u64,
+    ) -> EngineResult<SubmitResult> {
+        self.call(sender, receiver, value, Vec::new(), gas_limit)
+    }
+
+    /// Mint tokens for recipient on a particular ERC20 token
+    /// This function should return the amount of tokens unused,
+    /// which will be always all (<amount>) if there is any problem
+    /// with the input, or 0 if tokens were minted successfully.
+    ///
+    /// The output will be serialized as a String
+    /// https://github.com/near/NEPs/discussions/146
+    ///
+    /// IMPORTANT: This function should not panic, otherwise it won't
+    /// be possible to return the tokens to the sender.
+    pub fn receive_erc20_tokens(&mut self, args: &NEP141FtOnTransferArgs) {
+        let str_amount = args.amount.to_string();
+        let output_on_fail = str_amount.as_bytes();
```

Comment:
Any reason why we use this everywhere as the output on fail?
I wonder if the idea was to have 0 output on success and all other values signalize failure. What happens if the amount is zero?

Reply:
If the function fails for any reason we should return the full amount indicating the sender that this amount must be refunded. We return `0` on success indicating that nothing must be refunded.

>  What happens if the amount is zero?
In this case it is a void transfer, so either way (failure or success) nothing should be refunded.